document/core/exec/modules.rst

```diff
@@ -563,85 +563,95 @@ It is up to the :ref:`embedder <embedder>` to define how such conditions are rep
 
 4. For each :ref:`external value <syntax-externval>` :math:`\externval_i` in :math:`\externval^n` and :ref:`external type <syntax-externtype>` :math:`\externtype'_i` in :math:`\externtype_{\F{im}}^n`, do:
 
-   a. Assert: :math:`\externval_i` is :ref:`valid <valid-externval>` with :ref:`external type <syntax-externtype>` :math:`\externtype_i` in store :math:`S`.
+   a. If :math:`\externval_i` is not :ref:`valid <valid-externval>` with an :ref:`external type <syntax-externtype>` :math:`\externtype_i` in store :math:`S`, then:
+
+      i. Fail.
 
    b. If :math:`\externtype_i` does not :ref:`match <match-externtype>` :math:`\externtype'_i`, then:
 
       i. Fail.
 
-5. Let :math:`\moduleinst_{\F{im}}` be the auxiliary module instance :math:`\{\MIGLOBALS~\evglobals(\externval^\ast)\}` that only consists of the imported globals.
+.. _exec-initvals:
+
+5. Let :math:`\val^\ast` be the vector of :ref:`global <syntax-global>` initialization :ref:`values <syntax-val>` determined by :math:`\module` and :math:`\externval^n`. These may be calculated as follows.
+
+   a. Let :math:`\moduleinst_{\F{im}}` be the auxiliary module :ref:`instance <syntax-moduleinst>` :math:`\{\MIGLOBALS~\evglobals(\externval^n)\}` that only consists of the imported globals.
+
+   b. Let :math:`F_{\F{im}}` be the auxiliary :ref:`frame <syntax-frame>` :math:`\{ \AMODULE~\moduleinst_{\F{im}}, \ALOCALS~\epsilon \}`.
 
-6. Let :math:`F` be the :ref:`frame <syntax-frame>` :math:`\{ \AMODULE~\moduleinst_{\F{im}}, \ALOCALS~\epsilon \}`.
+   c. Push the frame :math:`F_{\F{im}}` to the stack.
 
-7. Push the frame :math:`F` to the stack.
+   d. For each :ref:`global <syntax-global>` :math:`\global_i` in :math:`\module.\MGLOBALS`, do:
 
-8. Let :math:`\globalidx_{\F{new}}` be the :ref:`global index <syntax-globalidx>` that corresponds to the number of global :ref:`imports <syntax-import>` in :math:`\module.\MIMPORTS` (i.e., the index of the first non-imported global).
+      i. :ref:`Execute <exec-expr>` the initializer expression :math:`\global_i.\GINIT`.
 
-9. For each :ref:`global <syntax-global>` :math:`\global_i` in :math:`\module.\MGLOBALS`, do:
+      ii. Assert: due to :ref:`validation <valid-global>`, an expression :math:`\val_i~\END` is now on the top of the stack.
```

Comment:
s/an expression/a value/   (the stack only has values)
s/~\END//  (I'd gloss over "executing" the `END` itself)

Okay, I now see why I had turned evaluation in its own judgement: it's to encapsulate the tedious repetition in the prose. I suggest that we keep the notion of evaluation that returns a value and the prose for it, but still just give the single-step formal rule -- it doesn't require an explicit rule for iterating that, since we can immediately express that with `^\ast`. A simple note might be enough. WDYT?

Reply:
Regarding evaluation, are you describing

	S; F; instr* ~~>* S'; F'; v
	-------------------------------
	S; F; instr* end ~~> S'; F'; v

If so, I agree. I liked the old evaluation judgment for the reason you give. I simply thought having `^\ast` part of the judgment's syntax was confusing.

Sanity check: Your `sed` script makes things a little less accurate but a little more clear, and will be moot if we return to an evaluation judgment.

Aside: I'd like to add and use an `\evals` macro that looks like `\stepto` but that links to this section. (Analogous to the various `\vdash*` macros.)

Reply:
My suggestion is to keep the change you did to the formal rule, but keep the old version of the textual rules. The idea being that in text we have to unfold the `^\ast`, but in the formal rules we don't need to because we can slap it on directly. IOW, "evaluation" is iterated expression execution.

Reply:
Ack.

Reply:
I addressed this and your other comments in cde59d8. (The changes at this point simply resurrect the old prose, now that there's a note about evaluation.)